### PR TITLE
Fix/eval scheduler actor wedge

### PR DESCRIPTION
**Pull request by ekirimlioglu**
Changes

  - Fixes a deterministic wedge in Controller.run_multi_pipeline_inference that caused multi-pipeline inference to hang at 0% GPU/CPU. Two independent triggers were closed:
    a. An exception during the batch-submission/bookkeeping block left the actor marked busy by PipelineScheduler.schedule() with no path to free it. The block is now wrapped in a try/except that mirrors the existing init-failure handler — marks the pipeline FAILED, drops the partial
  active_tasks entry, and calls scheduler.remove_pipeline so the actor is released.
    b. Actor death after dispatch (OOM, segfault) left orphaned futures and the controller's busy-branch was a blind time.sleep(0.5). Replaced with ray.wait(all_pending, num_returns=1, timeout=0.5) so failed futures surface and the existing reap path frees the actor on the next iteration.
  - No API or behavior changes for healthy runs. No breaking changes.

  Changelog Content

  Additions

  - Unit tests covering PipelineScheduler busy/free bookkeeping invariants, including a regression test for the actor-leak wedge (tests/test_pipeline_scheduler.py).

  Changes

  - Busy-loop in evals controller now blocks on ray.wait instead of time.sleep, so dead-actor futures are reaped promptly.

  Fixes

  - Prevent indefinite hang in evals multi-pipeline inference when batch dispatch raises synchronously or an actor dies after dispatch.

  Testing

  - I have added tests that prove my fix is effective or that my feature works
  - New and existing unit tests pass locally with my changes
  - I have tested this change manually
  - I have tested this change in the following environments:
    - Local development
    - Docker environment
    - Other: 10-config grid run that previously wedged on a multi-GPU host

  Screenshots (if applicable)

  N/A — backend-only change.

  Checklist

  - My code follows the project's style guidelines
  - I have performed a self-review of my own code
  - I have commented my code, particularly in hard-to-understand areas
  - I have made corresponding changes to the documentation (no user-facing docs affected)
  - My changes generate no new warnings
  - I have added tests that prove my fix is effective or that my feature works
  - New and existing unit tests pass locally with my changes
  - Any dependent changes have been merged and published

## Testing
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have tested this change manually
- [x] I have tested this change in the following environments:
  - [x] Local development
  - [ ] Docker environment
  - [ ] Other: _______________

## Screenshots (if applicable)
Add screenshots to help explain your changes.

## Checklist
- [ ] My code follows the project's style guidelines
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published

## Performance Impact
If this PR affects performance, describe the impact and any optimizations made.

## Related Issues
Fixes #(issue number)
Closes #(issue number)
Related to #(issue number)

---

> [!NOTE]
> **Medium Risk**
> Touches the multi-pipeline scheduling loop and Ray future waiting semantics; mistakes could cause new hangs or premature failures under load.
> 
> **Overview**
> Fixes a deterministic hang in `Controller.run_multi_pipeline_inference` by (1) wrapping batch dispatch/bookkeeping in a `try/except` that marks the pipeline FAILED and calls `scheduler.remove_pipeline()` so the actor is freed, and (2) replacing the busy-loop `time.sleep()` with a `ray.wait()` on *not-yet-ready* futures so dead-actor failures surface promptly.
> 
> Adds `tests/test_pipeline_scheduler.py` to codify scheduler bookkeeping invariants (busy actors must be freed via `set_completed_task` or `remove_pipeline`) and to cover the regression wedge scenario.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 1d33b88d4377c7cdb941eac844a3c5da8740e3ae. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: rapidfireai/evals/scheduling/controller.py [hunk 1]
```diff
@@ -1400,7 +1400,15 @@ def run_multi_pipeline_inference(
...
                         f"Gen: {status['current_generation']}"
                     )
-                time.sleep(0.5)
+                # Block on any in-flight future so a dead actor's failed futures
+                # surface here and get reaped on the next iteration.
+                all_pending = []
+                for task_info in active_tasks.values():
+                    all_pending.extend(task_info["futures"])
+                if all_pending:
+                    ray.wait(all_pending, num_returns=1, timeout=0.5)
```
[the remark this answers is not in the document]
Comment (ekirimlioglu) [reply rapidfireai/evals/scheduling/controller.py]: Fair point. Once any single batch future resolves on a healthy actor it permanently satisfies num_returns=1 and the 0.5s timeout never engages. With per-actor sequential batch processing and the reap path only draining when all of an actor's futures are ready, the all-actors-busy branch hot-spins through the IC poll and scheduler.schedule() for the rest of the shard.

 Fixed in this commit by partitioning first via a non-blocking ray.wait and only blocking on the not-yet-ready set:
 _ready, not_ready = ray.wait(all_futures, num_returns=len(all_futures), timeout=0)
  if not_ready:
     ray.wait(not_ready, num_returns=1, timeout=0.5)
 
 Dead-actor detection is preserved: a RayActorError-bearing future is "ready" to ray.wait, so it lands in _ready and gets drained by the next iteration's reap path (which calls ray.get and trips the existing except at controller.py:1312). When all in-flight futures are healthy and pending, we block on the unready set as intended; when some are already terminal, we fall through immediately and let the reap path act.

**Comment by arun-rfai** [comment]
@ekirimlioglu Thanks for the detailed diagnosis and patch! I verified that both failure modes are real and your fix follows the existing init-failure handler pattern, which is the right shape.

Before the RapidFire AI team can proceed with reviewing and merging this, please open a GitHub Issue first describing the bugs, following the repo's Issue conventions. Specifically:

- Repro steps (minimal sequence to trigger the wedge on main)
- Minimal set of configs that reliably reproduces it
- Machine environment (OS, Python version, Ray version, hardware — CPU/GPU, cores, memory)
- Workload characteristics (which model API calls, which embedding model, batch sizes, num actors, etc.)
- Observed symptoms (logs, stdout, how you confirmed the wedge vs. slow progress)
- Whether the wedge is deterministic or probabilistic on your setup
- Whether your patched branch resolved the wedge end-to-end on the original workload, or only addresses related failure modes

Then please reference that Issue in this PR as the one being resolved. This will help the QA team reproduce the bugs independently and validate the fix. It will also keep the project's bug-tracking conventions consistent.

Overall, thanks for the great contribution. First-time PRs of this depth are rare!

**Comment by ekirimlioglu** [comment]
Filed #239 with the repro, environment, and end-to-end validation notes.
 
Small suggestion: CONTRIBUTING.md doesn't currently mention the file-an-issue-first step. Adding a short note under "Submitting a pull request" could help future first-time contributors.

**Review by anay-rfai — CHANGES_REQUESTED** [review]
@ekirimlioglu Good catch and great work on the fix.

controller.py has gone through significant changes on main since you forked, most notably #237 (api gateway) and #243 (multimodal RAG). Could you update your branch against latest upstream/main before we merge?

1. Update your fork either with git merge or rebase against upstream/main — your call, both are fine. We'll squash-merge at the end.
2. Heads-up: the auto-merge will likely succeed without conflict, but please visually verify that your dispatch try/except still wraps the actor.process_batch.remote(...) loop and the active_tasks[actor_id] = {...} assignment in the new code shape. The surrounding regions were restructured.
3. Re-run tests against the updated branch to confirm the wedge repro still passes.

Once that's pushed, this is good to approve and a maintainer will squash-merge. Thanks!

**Comment by ekirimlioglu** [comment]
@anay-rfai 

Rebased. 5 commits replayed cleanly with no conflicts. Visually confirmed the dispatch try/except still wraps the process_batch.remote() loop, the active_tasks[actor_id] = {...} assignment, and the three db.set_actor_task_* writes (controller.py L1939-L1983 in the new shape). The except block still calls scheduler.remove_pipeline(pipeline_id) so the actor is freed on any synchronous failure during batch dispatch. The busy-loop ray.wait partition fix also survived at L1796-L1804.

Tests pass:

pytest tests/test_pipeline_scheduler.py  ->  7 passed in 0.96s

The regression test (test_actor_leaks_busy_when_neither_completion_nor_removal_called) still shows 20/20 busy-sentinel returns when the controller-side cleanup is skipped, confirming the wedge condition is still reachable in PipelineScheduler and the controller patch is what prevents

**Review by anay-rfai — APPROVED** [review]
LGTM
